Fix Snapshot Repository Corruption in Downgrade Scenarios

original-brownbear opened the pull request:
This PR introduces test infrastructure for downgrading a cluster while interacting with a given repository.
It fixes the fact that repository metadata in the new format could be written while there's still older snaphshots in the repository that require the old-format metadata to be restorable.

This PR does not yet introduce a new field in `RepositoryData` to avoid old versions from corrupting the repository. That will have to happen in a subsequent PR once the ability to read the new `min_version` field has been backported to `7.6` (see related TODO in the diff). 

Marked `non-issue` since this problem never made it into a release.

original-brownbear committed:
start

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into test-downgrade-bwc

original-brownbear committed:
works

original-brownbear committed:
works

original-brownbear committed:
works

original-brownbear committed:
reproduces at last

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into test-downgrade-bwc

original-brownbear committed:
fixed

original-brownbear committed:
nicer

original-brownbear committed:
add assertion

elasticmachine commented:
Pinging @elastic/es-distributed (:Distributed/Snapshot/Restore)

original-brownbear committed:
add min version

original-brownbear committed:
bck

original-brownbear committed:
fix compile

original-brownbear committed:
better efficiency

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into test-downgrade-bwc

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into test-downgrade-bwc

original-brownbear committed:
need 2 steps ...

original-brownbear committed:
document test

original-brownbear committed:
drier

original-brownbear committed:
drier

original-brownbear committed:
drier

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into test-downgrade-bwc

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into test-downgrade-bwc

original-brownbear committed:
nicer

ywelsch commented on qa/snapshot-repository-downgrade/src/test/java/org/elasticsearch/upgrades/ClusterDowngradeIT.java:
```diff
@@ -0,0 +1,220 @@
+
+    @Override
+    protected boolean preserveRollupJobsUponCompletion() {
+        return true;
```
do we need this?

ywelsch commented on qa/snapshot-repository-downgrade/src/test/java/org/elasticsearch/upgrades/ClusterDowngradeIT.java:
```diff
@@ -0,0 +1,220 @@
+
+    @Override
+    protected boolean preserveILMPoliciesUponCompletion() {
+        return true;
```
do we need this?

ywelsch commented on qa/snapshot-repository-downgrade/src/test/java/org/elasticsearch/upgrades/ClusterDowngradeIT.java:
```diff
@@ -0,0 +1,220 @@
+
+    @Override
+    protected boolean preserveSLMPoliciesUponCompletion() {
+        return true;
```
do we need this?

original-brownbear replied:
Nah I just blindly copy and pasted these, removed all but the repo and snapshot preserving now :)

ywelsch commented on qa/snapshot-repository-downgrade/src/test/java/org/elasticsearch/upgrades/ClusterDowngradeIT.java:
```diff
@@ -0,0 +1,220 @@
+            }
+            createSnapshot(client, repoName, "snapshot-" + TEST_STEP.toString().toLowerCase(Locale.ROOT));
+            final String snapshotToDeleteName = "snapshot-to-delete";
+            // Create a snapshot and delete it right away again to test the impact of each version's cleanup functionality that is run
```
good idea. In my steps above, account for this as well

ywelsch commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -351,6 +353,26 @@ public TimeValue timeout() {
     }
 
+    public boolean hasOldVersionSnapshots(String repositoryName, RepositoryData repositoryData, @Nullable SnapshotId excluded) {
+        final List<SnapshotId> snapshotIds = List.copyOf(repositoryData.getSnapshotIds());
```
why do we need to convert to a list here? Why is Collection not good enough=?

original-brownbear replied:
My bad, this was a left-over from when I used an API that needed a `List` to get all the snapshot versions before :)

ywelsch reviewed (changes requested):
Perhaps name the qa project just `repository-downgrade` (so that its name isn't so long)

original-brownbear committed:
CR: rename qa project

original-brownbear committed:
CR: remove dead code

original-brownbear committed:
CR: more tests and renaming

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into test-downgrade-bwc

original-brownbear committed:
rename qa projct

original-brownbear commented on qa/snapshot-repository-downgrade/src/test/java/org/elasticsearch/upgrades/ClusterDowngradeIT.java:
```diff
@@ -0,0 +1,220 @@
+        OLD,
+        UPGRADED,
+        DOWNGRADED,
+        RE_UPGRADED;
```
Yea that sounds nicer :)

original-brownbear commented:
@ywelsch thanks for taking a look! Addressed all points now and added the three different tests the way you suggested. This should be good for another round :)

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into test-downgrade-bwc

original-brownbear committed:
only use 2 clusters

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into test-downgrade-bwc

original-brownbear commented:
@ywelsch now it's using just two clusters of fixed version as discussed :)

ywelsch commented on qa/repository-multi-version/build.gradle:
```diff
@@ -0,0 +1,121 @@
+
+for (Version bwcVersion : bwcVersions.indexCompatible) {
+  String baseName = "v${bwcVersion}"
+  String revertedClusterName = "${baseName}-reverted"
```
reverted is a strange name. Perhaps use
```
String oldClusterName = "${baseName}-old"
String newClusterName = "${baseName}-new"
```

ywelsch commented on qa/repository-multi-version/src/test/java/org/elasticsearch/upgrades/MultiVersionRepositoryAccessIT.java:
```diff
@@ -0,0 +1,270 @@
+ * <ul>
+ *     <li>Start a cluster in an old version: {@link TestStep#STEP1_OLD_CLUSTER}</li>
+ *     <li>Start a cluster running the current version: {@link TestStep#STEP2_NEW_CLUSTER}</li>
+ *     <li>Again start a cluster in an old version: {@link TestStep#STEP3_OLD_CLUSTER}</li>
```
I'm confused by the terminology here. We're not starting a cluster again in step 3?

original-brownbear replied:
Reworded it a little, hope it's better now :)

ywelsch commented on qa/repository-multi-version/src/test/java/org/elasticsearch/upgrades/MultiVersionRepositoryAccessIT.java:
```diff
@@ -0,0 +1,270 @@
+
+    public void testCreateAndRestoreSnapshot() throws IOException {
+        final String repoName = "testCreateAndRestoreSnapshot";
+        try (RestHighLevelClient client = new RestHighLevelClient(RestClient.builder(adminClient().getNodes().toArray(new Node[0])))) {
```
why is it always taking the first node here?

original-brownbear replied:
It's taking all nodes? It's just creating the empty array for the conversion to `Node` not looking up `[]`:)

ywelsch replied:
Java 🗡

ywelsch commented on qa/repository-multi-version/src/test/java/org/elasticsearch/upgrades/MultiVersionRepositoryAccessIT.java:
```diff
@@ -0,0 +1,270 @@
+            // Create a snapshot and delete it right away again to test the impact of each version's cleanup functionality that is run
+            // as part of the snapshot delete
+            createSnapshot(client, repoName, snapshotToDeleteName);
+            deleteSnapshot(client, repoName, snapshotToDeleteName);
```
check that the snapshot can be listed before deletion?

original-brownbear replied:
Sure :)

ywelsch commented on qa/repository-multi-version/build.gradle:
```diff
@@ -0,0 +1,121 @@
+  def clusterSettings = { v ->
+    return {
+      version = v
+      numberOfNodes = 2
```
why 2 nodes? Isn't one node sufficient?

original-brownbear replied:
It's probably pointless paranoia but I figured if I add all these tests, why not also cover the over the wire serialization in this scenario as well (I can't see how it would be important but who knows ... could be we're tripping some assertion in a `Streamable` constructor or so ... we have beefy CI machines so why not? :)).

ywelsch commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -351,6 +354,26 @@ public TimeValue timeout() {
+                hasOldFormatSnapshots = false;
+            } else {
+                final Repository repository = repositoriesService.repository(repositoryName);
+                hasOldFormatSnapshots = snapshotIds.stream().map(repository::getSnapshotInfo).anyMatch(
```
what if we can't load one of these snapshot info? Should we be lenient in that case? Will this o.w. break snapshotting in a way that wasn't broken before?

original-brownbear replied:
This will brake snapshotting and deleting snapshots if a `snap-uuid` file doesn't exist.
I think it's much better to fail this case than to quietly ignore it. 
We simply should definitely not be making any updates to a corrupted repo ever IMO. This will not break restoring snapshots from the repo and users should simply move to a new repo when running into this kind of corruption. We discussed this before and I think the conclusion was that if a `snap-` file is missing ... what guarantees do we have some `__` data file isn't missing as well and quietly causing any new snapshot that uses it to be corrupted as well.
We also already eliminated a number of other spots around repo corruption like this where we were lenient before but aren't any longer.

ywelsch replied:
I'm mainly concerned about the situation in Cloud where repos that got previously broken due to eventual consistencies are not taking any new backups anymore, and require an operator to come in and fix the situation. I have a strong preference here to add special handling for the case where the meta file is not found and be lenient in that case (and assume it's an old style repo).

original-brownbear replied:
> I'm mainly concerned about the situation in Cloud where repos that got previously broken due to eventual consistencies are not taking any new backups anymore, and require an operator to come in and fix the situation.

I would argue that this is what we want. Assuming you have a snapshot in the `index-N` for which the `snap-` blob is missing at the repo root, then what guarantees do you have that the data blobs belonging to that snapshot and are referenced in the shard level `index-N` do actually exist and you're not just creating and endless stream of corrupted snapshots quietly? IMO this situation should be resolved in Cloud operationally (and if it turns out to be common, automated).

ywelsch replied:
> Assuming you have a snapshot in the index-N for which the snap- blob is missing at the repo root, then what guarantees do you have that the data blobs belonging to that snapshot and are referenced in the shard level index-N do actually exist and you're not just creating and endless stream of corrupted snapshots quietly?

I don't see how these two are related. Having a missing top-level `snap-` blob does not entail that you have missing segment blobs. I therefore don't follow the argument for endless corrupted snapshots. In any case, I don't think it's the responsibility of the snapshot creation / snapshot deletion API to verify unrelated snapshots, which is what this API now suddenly does with the changes here as it happens to need a full top-level scan of the repo. Let's leave this kind of verification to a separate API that checks integrity of the repo.

original-brownbear replied:
> I don't see how these two are related. Having a missing top-level snap- blob does not entail that you have missing segment blobs. I therefore don't follow the argument for endless corrupted snapshots.

Two points here:
1. The argument isn't specific to S3 here. The absence of a file that is supposed to be there but is not means that the repository is corrupted and other files could be missing for whatever reason (say a broken NFS setup).
2. Concretely to the point, If a `snap-` blob is missing due to eventual consistency issues the sequence of events would be:
  * `snap-` and all the other meta and data blobs related to a snapshot were deleted and a new index-N was written
  * Next operation failed to pick up `index-N` due to inconsistent listing and instead picked up `index-(N-1)` that still included the snapshot and then writes out `index-N` again still including the snapshot.
  * If the two operations were so close to each other, maybe we also picked up an older shard level `index-N` blob that references data blobs that are actually gone already? I don't think this can be ruled out at all and is somewhat likely even.

>  I don't think it's the responsibility of the snapshot creation / snapshot deletion API to verify unrelated snapshots, which is what this API now suddenly does with the changes here as it happens to need a full top-level scan of the repo. Let's leave this kind of verification to a separate API that checks integrity of the repo.

Agreed, I don't think it's the responsibility of snapshot create and delete to do any verification outright. But I though consensus was that if either of these operations detected the repository to be in a corrupted state, it wouldn't continue to make updates to the repo. If I remember correctly, the reasoning was precisely that if e.g. an NFS issue makes some files go missing and we continue to modify the repo ignoring that corruption, then we take away any chance of the operator to fix the storage layer instead of losing those snapshots for good?

original-brownbear replied:
@ywelsch as discussed on another channel, I adjusted things to now ignored a missing `snap-${uuid}.dat` blob and assume old generation in that case.
I also added a test for this scenario in 7d5d2565b0e (I know it's a little hacky, but I still liked it better than starting to manually mess with files in the repos used by the REST/BwC tests).

ywelsch reviewed (commented):
Nits and a serious comment.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into test-downgrade-bwc

original-brownbear committed:
CR: more checks + renaming

original-brownbear commented:
@ywelsch thanks, all addressed again I think

tlrx commented on qa/repository-multi-version/src/test/java/org/elasticsearch/upgrades/MultiVersionRepositoryAccessIT.java:
```diff
@@ -0,0 +1,272 @@
+
+/**
+ * Tests that verify that a snapshot repository is not getting corrupted and continues to function properly when accessed by multiple
+ * clusters of different versions. Concretely this test suit is simulating the following scenario:
```
suit -> suite

tlrx commented on qa/repository-multi-version/src/test/java/org/elasticsearch/upgrades/MultiVersionRepositoryAccessIT.java:
```diff
@@ -0,0 +1,272 @@
+    }
+
+    public void testCreateAndRestoreSnapshot() throws IOException {
+        final String repoName = "testCreateAndRestoreSnapshot";
```
Use getTestName()

original-brownbear replied:
neat :) didn't know about that one

tlrx commented on qa/repository-multi-version/src/test/java/org/elasticsearch/upgrades/MultiVersionRepositoryAccessIT.java:
```diff
@@ -0,0 +1,272 @@
+                snapshots.stream().map(sn -> (String) sn.get("snapshot")).toArray(String[]::new)), RequestOptions.DEFAULT);
+            for (SnapshotStatus status : statusResponse.getSnapshots()) {
+                assertThat(status.getShardsStats().getFailedShards(), is(0));
+            }
```
I think it deserves its own utility method, something like `assertSnapshotStatusSuccessful()`

original-brownbear replied:
Sounds good :) Done

tlrx commented on qa/repository-multi-version/src/test/java/org/elasticsearch/upgrades/MultiVersionRepositoryAccessIT.java:
```diff
@@ -0,0 +1,272 @@
+    }
+
+    private void deleteSnapshot(RestHighLevelClient client, String repoName, String name) throws IOException {
+        client.snapshot().delete(new DeleteSnapshotRequest(repoName, name), RequestOptions.DEFAULT);
```
Maybe check that the deletion is acked?

original-brownbear replied:
Sounds good :) Done

tlrx reviewed (approved):
LGTM - only left minor comments. I do like the tests.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into test-downgrade-bwc

original-brownbear committed:
nicer

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into test-downgrade-bwc

original-brownbear committed:
assume old version when missing snap- blob

original-brownbear commented:
Jenkins run elasticsearch-ci/2  (known unrelated ILM issue)

original-brownbear commented:
Jenkins run elasticsearch-ci/2 (known ILM test failure again)

ywelsch reviewed (approved):
LGTM. Thanks for the extra iterations!

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into test-downgrade-bwc

original-brownbear commented:
Thanks Yannick + Tanguy!
